src/basic/user-util.c

```diff
@@ -553,14 +550,7 @@ int getgroups_alloc(gid_t** gids) {
                 ngroups = getgroups(0, NULL);
                 if (ngroups < 0)
                         return -errno;
-                if (ngroups == 0)
-                        return false;
-
-                free(allocated);
-
-                p = allocated = new(gid_t, ngroups);
-                if (!allocated)
-                        return -ENOMEM;
+                free(p);
```

Comment:
And this basically means moving the whole definition of p into loop block...

Reply:
No, because we still have the last line which sets the output pointer: `*gids = TAKE_PTR(p);`:

```c
$ sed -n 529,560p src/basic/user-util.c
int getgroups_alloc(gid_t** gids) {
        _cleanup_free_  gid_t *p = NULL;
        int ngroups = 8;

        for (unsigned attempt = 0;;) {
                p = new(gid_t, ngroups);
                if (!p)
                        return -ENOMEM;

                ngroups = getgroups(ngroups, p);
                if (ngroups >= 0)
                        break;
                if (errno != EINVAL)
                        return -errno;

                /* Give up eventually */
                if (attempt++ > 10)
                        return -EINVAL;

                /* Get actual size needed, and size the array explicitly. Note that this is potentially racy
                 * to use (in multi-threaded programs), hence let's call this in a loop. */
                ngroups = getgroups(0, NULL);
                if (ngroups < 0)
                        return -errno;

                free(p);
        }

        *gids = TAKE_PTR(p);
        return ngroups;
}

```

Reply:
You could argue moving that line into the loop, but that just makes the loop fatter, and doesn't really improve anything.  It actually makes it less clear, since you have more code in a single block.  I think the current separation is nicer, as it shows clearly what happens on success.  You just need to see the last 2 lines of the function.

Reply:
BTW, you may be interested in changing the interface to the one I've designed for shadow utils:

```c
[[gnu::malloc(free)]]
gid_t *
agetgroups(size_t *ngids)
{
	int    n;
	gid_t  *gids;

	 ...

	*ngids = n;
	return gids;
}
```

By returning the allocated block instead of the number of groups, you can use the `[[gnu::malloc(free)]]` attribute, which gives it slightly more safety.  If you're interested in this change, please let me know, and I'll send a patch.

Reply:
in systemd we always want to return errors uniformly as negative errno in function returns – unless the really really only way a function can fail is OOM and nothing else. But that's not a given here, getgroups() can fail for a million reasons, and we want to be able to distinguish them without updating global variables.